jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java

```diff
@@ -0,0 +1,395 @@
+     */
+    public boolean release(Entry entry)
+    {
+        // first mark it as unused
+        boolean reusable = entry.tryRelease();
+
+        // then cache the released entry
+        if (cache != null && reusable)
+        {
+            List<Entry> cachedList = cache.get();
+            if (cachedList.size() < cacheSize)
+                cachedList.add(entry);
+        }
+        return reusable;
+    }
+
+    public boolean isClosed()
+    {
+        return closed;
+    }
+
+    @Override
+    public void close()
+    {
+        closeLock.lock();
+        try
+        {
+            closed = true;
+
+            // clear the shared list and close its entries
+            for (Entry entry : sharedList)
+            {
+                if (remove(entry) && entry instanceof Closeable)
+                {
+                    try
+                    {
+                        ((Closeable)entry).close();
+                    }
+                    catch (IOException e)
+                    {
+                        LOGGER.warn("Error closing entry {}", entry, e);
+                    }
+                }
+            }
+
+            // clear the thread-local caches
+            allCaches.forEach(List::clear);
+            allCaches.clear();
+        }
+        finally
+        {
+            closeLock.unlock();
+        }
+    }
+
+    /**
+     * Remove a value from the pool.
+     *
+     * @param entry the value to remove
+     * @return true if the entry was removed, false otherwise
+     */
+    public boolean remove(Entry entry)
+    {
+        if (!entry.tryRemove() && !closed)
+        {
+            if (LOGGER.isDebugEnabled())
+                LOGGER.debug("Attempt to remove an object from the pool that is still in use: {}", entry);
+            return false;
+        }
+
+        boolean removed = sharedList.remove(entry);
+        if (!removed && !closed)
+        {
+            if (LOGGER.isDebugEnabled())
+                LOGGER.debug("Attempt to remove an object from the pool that does not exist: {}", entry);
+        }
+
+        return removed;
+    }
+
+    public int size()
+    {
+        return sharedList.size();
+    }
+
+    public Collection<Entry> values()
+    {
+        return Collections.unmodifiableCollection(sharedList);
+    }
+
+    @Override
+    public String toString()
+    {
+        return getClass().getSimpleName() + " size=" + sharedList.size() + " entries=" + sharedList;
+    }
+
+    public class Entry
+    {
+        private final AtomicBiInteger state = new AtomicBiInteger(); // hi: 0=open,1=closed lo: multiplexing counter
+        private final AtomicInteger usageCount = new AtomicInteger();
+        private final T pooled;
+
+        public Entry(T pooled)
+        {
+            this.pooled = pooled;
+        }
+
+        public T getPooled()
+        {
+            return pooled;
+        }
+
+        public boolean tryAcquire()
+        {
+            if (maxUsageCount >= 0)
```

Comment:
`maxUsageCount` is volatile and should be copied in a local variable since it's used few lines below too.

Reply:
I don't believe it would matter, but one less volatile read is always a good thing so I'll do it anyways.